src/main/java/io/supertokens/bulkimport/BulkImportUserUtils.java

```diff
@@ -125,102 +123,198 @@ private static List<LoginMethod> getParsedLoginMethods(Main main, AppIdentifier
                 loginMethods.add(new LoginMethod(tenantId, recipeId, isVerified, isPrimary, timeJoinedInMSSinceEpoch, null, null, passwordlessLoginMethod));
-            } else if (recipeId != null) {
-                errors.add("Invalid recipeId for loginMethod. Pass one of emailpassword, thirdparty or, passwordless!");
             }
         }
         return loginMethods;
     }
 
-    private static void validateTenantId(Main main, AppIdentifier appIdentifier, String tenantId, String recipeId, List<String> errors)
+    private static String validateAndNormaliseUserRole(String role) {
+        // We just trim the role the CreateRoleAPI.java
+        return role.trim();
+    }
+
+    private static String validateAndNormaliseTotpSecretKey(String secretKey) {
+         // We don't perform any normalisation on the secretKey in ImportTotpDeviceAPI.java
+        return secretKey;
+    }
+
+    private static Integer validateAndNormaliseTotpPeriod(Number period, List<String> errors) {
+        // We don't perform any normalisation on the period in ImportTotpDeviceAPI.java other than checking if it is > 0
+        if (period != null && period.intValue() < 1) {
+            errors.add("period should be > 0 for a totp device.");
+            return null;
+        }
+        return period != null ? period.intValue() : null;
+    }
+
+    private static Integer validateAndNormaliseTotpSkew(Number skew, List<String> errors) {
+        // We don't perform any normalisation on the period in ImportTotpDeviceAPI.java other than checking if it is >= 0
+        if (skew != null && skew.intValue() < 0) {
+            errors.add("skew should be >= 0 for a totp device.");
+            return null;
+        }
+        return skew != null ? skew.intValue() : null;
+    }
+
+    private static String validateAndNormaliseTotpDeviceName(String deviceName) {
+        // We normalise the deviceName as per the ImportTotpDeviceAPI.java
+        return deviceName != null ? deviceName.trim() : null;
+    }
+
+    private static void validateAndNormaliseIsPrimaryField(JsonArray jsonLoginMethods, List<String> errors) {
+        // We are validating that only one loginMethod has isPrimary as true
+        boolean hasPrimaryLoginMethod = false;
+        for (JsonElement jsonLoginMethod : jsonLoginMethods) {
+            JsonObject jsonLoginMethodObj = jsonLoginMethod.getAsJsonObject();
+            if (validateJsonFieldType(jsonLoginMethodObj, "isPrimary", ValueType.BOOLEAN)) {
+                if (jsonLoginMethodObj.get("isPrimary").getAsBoolean()) {
+                    if (hasPrimaryLoginMethod) {
+                        errors.add("No two loginMethods can have isPrimary as true.");
+                    }
+                    hasPrimaryLoginMethod = true;
+                }
+            }
+        }
+    }
+
+    private static String validateAndNormaliseRecipeId(String recipeId, List<String> errors) {
+        if (recipeId == null) {
+            return null;
+        }
```

Comment:
this will never be null. So why have this?

Reply:
Same goes for a bunch of other validation functions

Reply:
It can be null during parsing but won't be null after we create the BulkImportUser